### PR TITLE
COR-267 fix(clerk-js): Add resetPassword method as a core resource

### PR DESCRIPTION
**Pull request by raptisj**
## Type of change

- [ ] 🐛 Bug fix
- [x] 🌟 New feature
- [ ] 🔨 Breaking change
- [ ] 📖 Refactoring / dependency upgrade / documentation
- [ ] other:

## Packages affected

- [x] `@clerk/clerk-js`
- [ ] `@clerk/clerk-react`
- [ ] `@clerk/nextjs`
- [ ] `@clerk/remix`
- [x] `@clerk/types`
- [ ] `@clerk/themes`
- [ ] `@clerk/localizations`
- [ ] `@clerk/clerk-expo`
- [ ] `@clerk/backend`
- [ ] `@clerk/clerk-sdk-node`
- [ ] `@clerk/shared`
- [ ] `@clerk/fastify`
- [ ] `@clerk/chrome-extension`
- [ ] `gatsby-plugin-clerk`
- [ ] `build/tooling/chore`

## Description

- [x] `npm test` runs as expected.
- [x] `npm run build` runs as expected.

Add `resetPassword` method as a core resource and add `reset_password_code` strategy in prepare_first_factors.

### CODE REVIEW

**Review by panteliselef — CHANGES_REQUESTED** [review]

**Comment by panteliselef** [comment]
@raptisj  Should we add a few tests for the `resetPassword` method + the new cases for prepareFirstFactor ?

File: packages/clerk-js/src/core/resources/SignIn.ts [hunk 1]
```diff
@@ -61,6 +63,13 @@ export class SignIn extends BaseResource implements SignInResource {
     });
   };
 
+  resetPassword = (params: ResetPasswordParams): Promise<SignInResource> => {
+    return this._basePost({
+      body: params,
+      action: 'prepare_first_factor',
```
Comment (panteliselef) [inline packages/clerk-js/src/core/resources/SignIn.ts:69]: ⛔️ ```suggestion
      action: 'reset_password',
```

**Review by panteliselef — CHANGES_REQUESTED** [review]

File: packages/types/src/factors.ts [hunk 2]
```diff
@@ -53,6 +54,15 @@ export type BackupCodeFactor = {
   strategy: BackupCodeStrategy;
 };
 
+export type ResetPasswordCodeFactor = {
+  strategy: ResetPasswordCodeStrategy;
+  emailAddressId?: string;
+  safeIdentifier: string;
+  phoneNumberId?: string;
```
Comment (dimkl) [inline packages/types/src/factors.ts:61]: ❓ Can both `emailAddressId` and `phoneNumberId` be present? If not, can we make the type mutually exclusive?

File: packages/types/src/factors.ts [hunk 3]
```diff
@@ -98,3 +108,8 @@ export type BackupCodeAttempt = {
   strategy: BackupCodeStrategy;
   code: string;
 };
+
+export type ResetPasswordCodeAttempt = {
```
Comment (dimkl) [inline packages/types/src/factors.ts:112]: ❓ i cannot find a reference in the PR. Is this used somewhere else?

File: packages/types/src/signIn.ts [hunk 4]
```diff
@@ -79,7 +82,12 @@ export interface SignInResource extends ClerkResource {
   createMagicLinkFlow: () => CreateMagicLinkFlowReturn<SignInStartMagicLinkFlowParams, SignInResource>;
 }
 
-export type SignInStatus = 'needs_identifier' | 'needs_first_factor' | 'needs_second_factor' | 'complete';
+export type SignInStatus =
+  | 'needs_identifier'
+  | 'needs_first_factor'
+  | 'needs_second_factor'
+  | 'needs_new_password'
```
Comment (dimkl) [inline packages/types/src/signIn.ts:89]: ❓ could someone add it also in the FAPI openapi ?
https://github.com/clerkinc/clerk_go/blob/2c67365e804fc6cb651661c034561ae2da5d1b1e/api/client/openapi/schemas/Client.yml#L979

File: packages/types/src/factors.ts (hunk 3 above)
Comment (panteliselef) [reply packages/types/src/factors.ts:112]: It is being used by this [PR](https://github.com/clerkinc/javascript/pull/1079/files)

File: packages/types/src/factors.ts (hunk 2 above)
Comment (panteliselef) [reply packages/types/src/factors.ts:61]: We decided that `phoneNumberId`  and `emailAddressId` will no longer be needed, and we are removing them. A [PR](https://github.com/clerkinc/javascript/pull/1079/files) is already open for this

**Pull request merged**